network-api/networkapi/buyersguide/management/commands/migrate_products.py

```diff
@@ -0,0 +1,237 @@
+import ntpath
+
+from django.conf import settings
+from django.core.files.images import ImageFile
+from django.core.management.base import BaseCommand
+from io import BytesIO
+from mimetypes import MimeTypes
+from PIL import Image as PILImage
+
+from wagtail.images.models import Image as WagtailImage
+
+from networkapi.buyersguide.pagemodels.products.base import Product
+from networkapi.buyersguide.pagemodels.products.general import GeneralProduct
+from networkapi.buyersguide.pagemodels.products.software import SoftwareProduct
+from networkapi.wagtailpages.pagemodels.base import Homepage
+from networkapi.wagtailpages.pagemodels.products import (
+    ProductPagePrivacyPolicyLink, ProductPage, SoftwareProductPage,
+    GeneralProductPage, ProductPageCategory, BuyersGuidePage,
+    RelatedProducts, ProductUpdates
+)
+
+
+class Command(BaseCommand):
+    help = 'Migrate PNI Product models to Wagtail ProductPage types'
+
+    # Get or create the buyersguidepage
+    def get_or_create_buyers_guide(self):
+        """
+        Return the first BuyersGuidePage, or create a new one.
+        """
+        buyersguide_page = BuyersGuidePage.objects.first()
+        if not buyersguide_page:
+            # Create the buyersguide page.
+            buyersguide_page = BuyersGuidePage()
+            buyersguide_page.title = 'Privacy not included'
+            buyersguide_page.slug = 'privacynotincluded'
+            buyersguide_page.slug_en = 'privacynotincluded'
+            homepage = Homepage.objects.first()
+            homepage.add_child(instance=buyersguide_page)
+            buyersguide_page.save_revision().publish()
+        return buyersguide_page
+
+    def debug_print(self, *messages):
+        """For local debugging and testing only."""
+        if settings.DEBUG:
+            message = ' '.join([str(x) for x in messages])
+            print(message)
```

Comment:
For locally running this script to see what's being executed. It's not needed, honestly, but it's a handy little method to have around when running large migration scripts on localhost for faster debugging and accuracy. 

Typically I'd toss this out when opening a PR, but seeing how this script should be short lived in production I didnt see the harm in being here while we write tests

Reply:
I did notice this can interfere with stack tracing, where it'll print literally in the middle of a stack print O_o
```
Creating foundationmozillaorg_backend_run ... done
Using a GeneralProductPage
        Setting fields:
                 slug  as  google-nest-hub-max
                 privacy_ding  as  False
                 adult_content  as  False
                 uses_wifi  as  True
                 uses_bluetooth  as  True
                 review_date  as  2020-11-02
                 company  as  Google
                 blurb  as  Meet Google's high-end home hub complete with microphone, Nest Cam camera, and touchscreen display. Use it for video calls, monitoring your video doorbell, watching 
YouTube, sharing photos, controlling your connected thermostat, listening to music, or getting alerts whenever the camera's motion sensor picks up something moving in your home. It's a good thing Google built a physical way to turn off the camera and microphone on this gadget because it could be a better spy in your home than James Bond.
                 price  as  229
                 worst_case  as  Google seems to do a better job than some of the other Big Tech companies when it comes to privacy (looking at you Facebook). We like that users are now automatically opted out of Google's human review of voice recordings. However, Google does collect a ton of data on you, especially if you don't take the time to adjust your privacy setting to lock down just how much info they can gather. We recognize it is time consuming to have to go in and adjust setting in apps and on your phone and on your computer to protect your privacy. If you don't take the time, it's possible Google can get to know you really well, maybe too well. Maybe they recognize you from all the times you ordered plain cheese pizza. They know you are single because who orders plain cheese pizza? Just kidding, they know you're single because of all those pedicure appointments you've booked for one. Maybe it's OK Google knows you so well. Maybe it's creepy.
                 signup_requires_email  as  Yes
                 signup_requires_phone  as  No
                 signup_requires_third_party_account  as  No
                 signup_requirement_explanation  as
                 how_does_it_use_data_collected  as  While video recordings, voice recordings, and home environment sensor readings won't be used for for ad personalization, transcripts of your voice interactions may be. Non-personally identifiable information may be shared with partners. Google says that your personal data is never sold to third parties.
                 data_collection_policy_is_bad  as  False
                 user_friendly_privacy_policy  as  Yes
                 show_ding_for_minimum_security_standards  as  False
                 meets_minimum_security_standards  as  True
                 uses_encryption  as  No
                 uses_encryption_helptext  as  Uses encryption in transit and at rest.
                 security_updates  as  Yes
                 security_updates_helptext  as
                 strong_password  as  Yes
                 strong_password_helptext  as
                 manage_vulnerabilities  as  Yes
                 manage_vulnerabilities_helptext  as  Google has a STraceback (most recent call last):
  File "network-api/manage.py", line 34, in <module>
    execute_from_command_line(sys.argv)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/base.py", line 323, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/base.py", line 364, in execute
    output = self.handle(*args, **options)
  File "/app/network-api/networkapi/buyersguide/management/commands/migrate_products.py", line 133, in handle
    mime_type = mime.guess_type(product.image.file.name)  # -> ('image/jpeg', None)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/db/mecurity Rewards program, which means that anyone who finds a security issue and discloses it responsibly may get paid. https://www.google.com/about/appsecurity/programs-home/
                 privacy_policy  as  Yes
                 privacy_policy_helptext  as  All things Nest have been moved over to the Google Store. Google has a page detailing their privacy principles for Nest products.
                 phone_number  as  (855) â€‘469â€‘6378
                 live_chat  as  https://nest.secure.force.com/prechatform/apex/PreChatForm?language=en_US&caseNumber=&endpoint=https
                 email  as
                 twitter  as  googlenest
                 camera_device  as  Yes
                 camera_app  as  Yes
                 microphone_device  as  Yes
                 microphone_app  as  Yes
                 location_device  as  Yes
                 location_app  as  Yes
                 personal_data_collected  as  Name, email, phone number
                 biometric_data_collected  as  Audio recordings, face match
                 social_data_collected  as  Contacts
                 how_can_you_control_your_data  as  You can review and delete your data in the Nest app or via the Google Privacy odels/fields/files.py", line 41, in _get_file
    self._require_file()
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/db/models/fields/files.py", line 38, in _require_file
    raise ValueError("The '%s' attribute has no file associated with it." % self.field.name)
ValueError: The 'image' attribute has no file associated with it.
Dashboard.
                 data_control_policy_is_bad  as  False
                 track_record_choices  as  [('Great', 'Great'), ('Average', 'Average'), ('Needs Improvement', 'Needs Improvement'), ('Bad', 'Bad')]
                 company_track_record  as  Needs Improvement
                 track_record_is_bad  as  True
                 track_record_details  as  In 2018, personal information was exposed in several incidents, resulting from access by app developers, exposure by mobile apps, or a bug in Google + API. Google has also stored location data without permission.
                 offline_capable  as  No
                 offline_use_description  as
                 uses_ai  as  Yes
                 ai_uses_personal_data  as  Yes
                 ai_is_transparent  as  Yes
                 ai_helptext  as  Google publishes academic papers about it's AI research (https://ai.google/) and makes several tools available via open source. https://ai.google/tools/ Learn 
more about familiar face detection: https://support.google.com/googlenest/answer/9268625?co=GENIE.Platform%3DAndroid&hl=en
```

Reply:
Interesting. How'd you get this to print in the middle of a stack trace? Mine either falls over with a big ol' stack trace at the end or prints what it should print. 🤔

Reply:
"I just ran the code" =/

Reply:
I reran `new-env`, and this time it worked, so I'm going to try this on real data again, using:

1. `inv new-env` while on master
1. pull prod db as local db
1. `inv checkout 5520-product-model-to-page-migration-script`
1. `inv migrate`
1. `inv manage migrate_products`

Reply:
Yeah, same error:
```
Starting foundationmozillaorg_postgres_1 ... done
Creating foundationmozillaorg_backend_run ... done
Using a GeneralProductPage
        Setting fields:
                 slug  as  google-nest-hub-max
                 privacy_ding  as  False      
                 adult_content  as  False     
                 uses_wifi  as  True
                 uses_bluetooth  as  True     
                 review_date  as  2020-11-02  
                 company  as  Google
                 blurb  as  Meet Google's high-end home hub complete with microphone, Nest Cam camera, and touchscreen display. Use it for video calls, monitoring your video doorbell, watching YouTube, sharing photos, controlling your connected thermostat, listening to music, or getting alerts whenever 
the camera's motion sensor picks up something moving in your home. It's a good thing Google built a physical way to turn off the camera and microphone on this gadget because it could be a better spy in your home than James Bond.
                 price  as  229
                 worst_case  as  Google seems to do a better job than some of the other Big Tech companies when it comes to privacy (looking at you Facebook). We like that users are now automatically opted out of Google's human review of voice recordings. However, Google does collect a ton of data on you, especially if you don't take the time to adjust your privacy setting to lock down just how much info they can gather. We recognize it is time consuming to have to go in and adjust setting in apps and on your phone and on your computer to protect your privacy. If you don't take the time, it's possible Google can get to know you really well, maybe too well. Maybe they recognize you from all the times you ordered plain cheese pizza. They know you are single because who orders plain cheese pizza? Just kidding, they know you're single because of all those pedicure appointments you've booked for one. Maybe it's OK Google knows you so well. Maybe it's creepy.
                 signup_requires_email  as  Yes
                 signup_reTraceback (most recent call last):
quires_phone  as  No
                 signup_requires_third_party_account  as  No
                 signup_requirement_explanation  as
                 how_does_it_use_data_collected  as  While video recordings, voice recordings, and home environment sensor readings won't be used for ad personalization, transcripts of your voice interactions may be. Non-personally identifiable information may be shared with partners. Google says that your personal data is never sold to third parties.
                 data_collection_policy_is_bad  as  False
                 user_friendly_privacy_policy  as  Yes
                 show_ding_for_minimum_security_standards  as  False
                 meets_minimum_security_standards  as  True
                 uses_encryption  as  No
                 uses_encryption_helptext  as  Uses encryption in transit and at rest.
                 security_updates  as  Yes
                 security_updates_helptext  as
                 strong_password  as  Yes
                 strong_password_helptext  as
                 manage_vulnerabilities  as  Yes
                 manage_vulnerabilities_helptext  as  Google has a Security Rewards program, which means that anyone wh  File "network-api/manage.py", line 34, in <module>
    execute_from_command_line(sys.argv)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/base.py", line 323, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/base.py", line 364, in execute
    output = self.handle(*args, **options)
  File "/app/network-api/networkapi/buyersguide/management/commands/migrate_products.py", line 133, in handle
    mime_type = mime.guess_type(product.image.file.name)  # -> ('image/jpeg', None)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/db/mo finds a security issue and discloses it responsibly may get paid. https://www.google.com/about/appsecurity/programs-home/
                 privacy_policy  as  Yes
                 privacy_policy_helptext  as  All things Nest have been moved over to the Google Store. Google has a page detailing their privacy principles for Nest products.
                 phone_number  as  (855) â€‘469â€‘6378
                 live_chat  as  https://nest.secure.force.com/prechatform/apex/PreChatForm?language=en_US&caseNumber=&endpoint=https
                 email  as
                 twitter  as  googlenest
                 camera_device  as  Yes
                 camera_app  as  Yes
                 microphone_device  as  Yes
                 microphone_app  as  Yes
                 location_device  as  Yes
                 location_app  as  Yes
                 personal_data_collected  as  Name, email, phone number
                 biometric_data_collected  as  Audio recordings, face match
                 social_data_collected  as  Contacts
                 how_can_you_control_your_data  as  You can review and delete your data in the Nest app or via the Google Privacy Dashboard.
                 data_control_policy_is_bad  as  Falseodels/fields/files.py", line 41, in _get_file
    self._require_file()
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/db/models/fields/files.py", line 38, in _require_file
    raise ValueError("The '%s' attribute has no file associated with it." % self.field.name)
ValueError: The 'image' attribute has no file associated with it.

                 track_record_choices  as  [('Great', 'Great'), ('Average', 'Average'), ('Needs Improvement', 'Needs Improvement'), ('Bad', 'Bad')]
                 company_track_record  as  Needs Improvement
                 track_record_is_bad  as  True
                 track_record_details  as  In 2018, personal information was exposed in several incidents, resulting from access by app developers, exposure by mobile apps, or a bug in Google + API. Google has also stored location data without permission.
                 offline_capable  as  No
                 offline_use_description  as
                 uses_ai  as  Yes
                 ai_uses_personal_data  as  Yes
                 ai_is_transparent  as  Yes
                 ai_helptext  as  Google publishes academic papers about it's AI research (https://ai.google/) and makes several tools available via open source. https://ai.google/tools/ Learn more about familiar face detection: https://support.google.com/googlenest/answer/9268625?co=GENIE.Platform%3DAndroid&hl=en
```

Reply:
no more errors!